File: examples/js/src/standalone/schelling-study.ts
// Reusable scientific loop shared by the user CLI and publication adapter.
// The extraction prevents drift; it is not required by the JS binding.
import {
  DEFAULT_SCHELLING_CONFIG,
  SchellingModel,
  type SchellingConfig,
} from '../models/schelling';

export type SchellingStudyMode = 'steady' | 'convergence';

export interface SchellingStudyOptions {
  width: number;
  height: number;
  density: number;
  balance: number;
  thresholds: number[];
  warmupSteps: number;
  steps: number;
  seeds: number;
  seed: number;
  mode: SchellingStudyMode;
}

export interface SchellingTrialResult {
  readonly threshold: number;
  readonly seed: number;
  readonly elapsedMs: number;
  readonly satisfiedPct: number;
  readonly segregationIndex: number;
  readonly stepsRun: number;
  readonly converged: boolean;
  readonly lastMoved: number;
}

export interface SchellingStudyRow {
  readonly threshold: number;
  readonly meanSatisfiedPct: number;
  readonly meanSegregationIndex: number;
  readonly meanLastMoved: number;
  readonly meanSteps: number;
  readonly convergedRuns: number;
}

export interface SchellingStudyResult {
  readonly options: SchellingStudyOptions;
  readonly trials: readonly SchellingTrialResult[];
  readonly rows: readonly SchellingStudyRow[];
  readonly elapsedMs: number;
  readonly totalTicks: number;
  readonly msPerTick: number;
  readonly satisfiedPct: number;
  readonly segregationIndex: number;
  readonly actualSteps: number;
}

export const DEFAULT_SCHELLING_STUDY_OPTIONS: SchellingStudyOptions = {
  width: DEFAULT_SCHELLING_CONFIG.gridWidth,
  height: DEFAULT_SCHELLING_CONFIG.gridHeight,
  density: DEFAULT_SCHELLING_CONFIG.density,
  balance: DEFAULT_SCHELLING_CONFIG.balance,
  thresholds: [0.3, 0.5, 0.7, 0.9],
  warmupSteps: 0,
  steps: 1_000,
  seeds: 8,
  seed: 7,
  mode: 'convergence',
};

function optionValues(argv: readonly string[]): Map<string, string | true> {
  const values = new Map<string, string | true>();
  const allowed = new Set([
    'width', 'height', 'density', 'balance', 'thresholds',
    'warmup-steps', 'steps', 'seeds', 'seed', 'mode',
  ]);
  for (let index = 0; index < argv.length; index += 1) {
    const argument = argv[index]!;
    if (!argument.startsWith('--')) throw new Error(`Unexpected argument: ${argument}`);
    const equalsAt = argument.indexOf('=');
    const name = argument.slice(2, equalsAt >= 0 ? equalsAt : undefined);
    if (!allowed.has(name)) throw new Error(`Unknown option: --${name}`);
    if (equalsAt >= 0) {
      values.set(name, argument.slice(equalsAt + 1));
      continue;
    }
    const next = argv[index + 1];
    if (next === undefined || next.startsWith('--')) throw new Error(`--${name} requires a value.`);
    values.set(name, next);
    index += 1;
  }
  return values;
}

function finiteNumber(values: Map<string, string | true>, name: string, fallback: number): number {
  const raw = values.get(name);
  if (raw === undefined) return fallback;
  if (raw === true) throw new Error(`--${name} requires a number.`);
  const value = Number(raw);
  if (!Number.isFinite(value)) throw new Error(`--${name} must be a finite number.`);
  return value;
}

function positiveInteger(values: Map<string, string | true>, name: string, fallback: number): number {
  const value = finiteNumber(values, name, fallback);
  if (!Number.isInteger(value) || value <= 0) throw new Error(`--${name} must be a positive integer.`);
  return value;
}

function nonNegativeInteger(values: Map<string, string | true>, name: string, fallback: number): number {
  const value = finiteNumber(values, name, fallback);
  if (!Number.isInteger(value) || value < 0) throw new Error(`--${name} must be a non-negative integer.`);
  return value;
}

export function parseSchellingThresholds(value: string): number[] {
  const thresholds = value.split(',').filter(Boolean).map((item) => Number(item.trim()));
  if (thresholds.length === 0 || thresholds.some((threshold) => !Number.isFinite(threshold) || threshold < 0 || threshold > 1)) {
    throw new Error('--thresholds must contain one or more values from 0 through 1.');
  }
  return thresholds;
}

export function parseSchellingStudyOptions(
  argv: readonly string[],
  defaults: SchellingStudyOptions = DEFAULT_SCHELLING_STUDY_OPTIONS,
): SchellingStudyOptions {
  const values = optionValues(argv);
  const mode = values.get('mode') ?? defaults.mode;
  if (mode !== 'steady' && mode !== 'convergence') throw new Error('--mode must be steady or convergence.');
  const density = finiteNumber(values, 'density', defaults.density);
  const balance = finiteNumber(values, 'balance', defaults.balance);
  if (density < 0 || density > 1 || balance < 0 || balance > 1) {
    throw new Error('--density and --balance must be values from 0 through 1.');
  }
  const thresholdsRaw = values.get('thresholds');
  return {
    width: positiveInteger(values, 'width', defaults.width),
    height: positiveInteger(values, 'height', defaults.height),
    density,
    balance,
    thresholds: thresholdsRaw === undefined
      ? [...defaults.thresholds]
      : parseSchellingThresholds(thresholdsRaw as string),
    warmupSteps: nonNegativeInteger(values, 'warmup-steps', defaults.warmupSteps),
    steps: positiveInteger(values, 'steps', defaults.steps),
    seeds: positiveInteger(values, 'seeds', defaults.seeds),
    seed: Math.trunc(finiteNumber(values, 'seed', defaults.seed)),
    mode,
  };
}

export function runSchellingTrial(
  config: SchellingConfig,
  steps: number,
  mode: SchellingStudyMode,
): Omit<SchellingTrialResult, 'threshold' | 'seed'> {
  const model = new SchellingModel(config);
  model.initialize();
  const started = performance.now();
  let stepsRun = 0;
  let converged = false;
  for (let step = 0; step < steps; step += 1) {
    stepsRun += 1;
    const moved = model.step();
    if (mode === 'convergence' && !moved) {
      converged = true;
      break;
    }
  }
  const elapsedMs = performance.now() - started;
  const statistics = model.getStatistics();
  return {
    elapsedMs,
    satisfiedPct: statistics.satisfactionRate,
    segregationIndex: statistics.segregationIndex,
    stepsRun,
    converged,
    lastMoved: statistics.lastMoved,
  };
}

export function runSchellingStudy(options: SchellingStudyOptions): SchellingStudyResult {
  const baseConfig = {
    gridWidth: options.width,
    gridHeight: options.height,
    density: options.density,
    balance: options.balance,
  };
  if (options.warmupSteps > 0) {
    runSchellingTrial({
      ...baseConfig,
      similarityThreshold: options.thresholds[0]!,
      seed: options.seed,
    }, options.warmupSteps, 'steady');
  }

  const trials: SchellingTrialResult[] = [];
  for (const threshold of options.thresholds) {
    for (let seedOffset = 0; seedOffset < options.seeds; seedOffset += 1) {
      const seed = options.seed + seedOffset;
      trials.push({
        threshold,
        seed,
        ...runSchellingTrial({ ...baseConfig, similarityThreshold: threshold, seed }, options.steps, options.mode),
      });
    }
  }

  const elapsedMs = trials.reduce((total, trial) => total + trial.elapsedMs, 0);
  const totalTicks = trials.reduce((total, trial) => total + trial.stepsRun, 0);
  const trialCount = trials.length;
  const rows = options.thresholds.map((threshold) => {
    const selected = trials.filter((trial) => trial.threshold === threshold);
    return {
      threshold,
      meanSatisfiedPct: selected.reduce((total, trial) => total + trial.satisfiedPct, 0) / selected.length,
      meanSegregationIndex: selected.reduce((total, trial) => total + trial.segregationIndex, 0) / selected.length,
      meanLastMoved: selected.reduce((total, trial) => total + trial.lastMoved, 0) / selected.length,
      meanSteps: selected.reduce((total, trial) => total + trial.stepsRun, 0) / selected.length,
      convergedRuns: selected.filter((trial) => trial.converged).length,
    };
  });
  return {
    options,
    trials,
    rows,
    elapsedMs,
    totalTicks,
    msPerTick: totalTicks === 0 ? 0 : elapsedMs / totalTicks,
    satisfiedPct: trials.reduce((total, trial) => total + trial.satisfiedPct, 0) / trialCount,
    segregationIndex: trials.reduce((total, trial) => total + trial.segregationIndex, 0) / trialCount,
    actualSteps: totalTicks / trialCount,
  };
}

export function formatSchellingStudyCsv(result: SchellingStudyResult): string {
  const lines = ['threshold,mean_satisfied_pct,mean_segregation_index,mean_last_swapped,mean_steps,converged_runs'];
  for (const row of result.rows) {
    lines.push([
      row.threshold.toFixed(2),
      row.meanSatisfiedPct.toFixed(4),
      row.meanSegregationIndex.toFixed(4),
      row.meanLastMoved.toFixed(2),
      row.meanSteps.toFixed(2),
      row.convergedRuns,
    ].join(','));
  }
  lines.push(
    'performance_metric,total_ticks,elapsed_ms,tpms,mspt',
    [
      'performance', result.totalTicks, result.elapsedMs.toFixed(3),
      (result.elapsedMs === 0 ? 0 : result.totalTicks / result.elapsedMs).toFixed(6),
      result.msPerTick.toFixed(6),
    ].join(','),
  );
  return `${lines.join('\n')}\n`;
}

File: examples/julia/schelling_study.jl
# Reusable scientific loop shared by the user CLI and publication kernel. The
# extraction prevents drift; it is not required by Agents.jl or TenSnap.
using Printf
using Statistics

Base.@kwdef struct SchellingStudyOptions
	width::Int = DEFAULT_GRID_W
	height::Int = DEFAULT_GRID_H
	density::Float64 = DEFAULT_DENSITY
	balance::Float64 = DEFAULT_BALANCE
	thresholds::Vector{Float64} = [0.3, 0.5, 0.7, 0.9]
	warmup_steps::Int = 0
	steps::Int = 1000
	seeds::Int = 8
	seed::Int = 7
	mode::Symbol = :convergence
end

function option_values(args)::Dict{String, String}
	allowed = Set(["width", "height", "density", "balance", "thresholds", "warmup-steps", "steps", "seeds", "seed", "mode"])
	values = Dict{String, String}()
	index = 1
	while index <= length(args)
		argument = args[index]
		startswith(argument, "--") || error("unexpected argument: $argument")
		body = argument[3:end]
		if occursin("=", body)
			name, value = split(body, "="; limit = 2)
		else
			name = body
			index < length(args) || error("--$name requires a value")
			value = args[index + 1]
			startswith(value, "--") && error("--$name requires a value")
			index += 1
		end
		name in allowed || error("unknown option: --$name")
		values[name] = value
		index += 1
	end
	return values
end

function parse_thresholds(raw::String)::Vector{Float64}
	thresholds = [parse(Float64, strip(item)) for item in split(raw, ",") if !isempty(strip(item))]
	!isempty(thresholds) || error("at least one threshold is required")
	all(value -> 0 <= value <= 1, thresholds) || error("thresholds must be values from 0 through 1")
	return thresholds
end

function parse_schelling_study_options(args = ARGS)::SchellingStudyOptions
	values = option_values(args)
	value(name, fallback) = get(values, name, string(fallback))
	mode = Symbol(value("mode", "convergence"))
	mode in (:steady, :convergence) || error("mode must be steady or convergence")
	options = SchellingStudyOptions(
		width = parse(Int, value("width", DEFAULT_GRID_W)),
		height = parse(Int, value("height", DEFAULT_GRID_H)),
		density = parse(Float64, value("density", DEFAULT_DENSITY)),
		balance = parse(Float64, value("balance", DEFAULT_BALANCE)),
		thresholds = parse_thresholds(value("thresholds", "0.30,0.50,0.70,0.90")),
		warmup_steps = parse(Int, value("warmup-steps", 0)),
		steps = parse(Int, value("steps", 1000)),
		seeds = parse(Int, value("seeds", 8)),
		seed = parse(Int, value("seed", 7)),
		mode = mode,
	)
	options.width > 0 && options.height > 0 || error("width and height must be positive")
	0 <= options.density <= 1 && 0 <= options.balance <= 1 || error("density and balance must be values from 0 through 1")
	options.steps > 0 && options.seeds > 0 || error("steps and seeds must be positive")
	options.warmup_steps >= 0 || error("warmup-steps must be non-negative")
	return options
end

function run_schelling_trial(config::SchellingConfig, steps::Int, mode::Symbol)
	model = initialize_schelling(config)
	steps_run = 0
	converged = false
	started_ns = time_ns()
	for _ in 1:steps
		steps_run += 1
		moving = schelling_model_step!(model)
		if mode == :convergence && !moving
			converged = true
			break
		end
	end
	return (
		threshold = config.similarity_threshold,
		seed = config.seed,
		satisfied = satisfied_pct(model),
		segregation = segregation_index(model),
		last_swapped = Agents.abmproperties(model).last_swapped,
		steps_run = steps_run,
		converged = converged,
		elapsed_ns = time_ns() - started_ns,
	)
end

function run_schelling_study(options::SchellingStudyOptions)
	config_for(threshold, seed) = SchellingConfig(
		gridwidth = options.width,
		gridheight = options.height,
		density = options.density,
		balance = options.balance,
		similarity_threshold = threshold,
		seed = seed,
	)
	if options.warmup_steps > 0
		run_schelling_trial(config_for(options.thresholds[1], options.seed), options.warmup_steps, :steady)
	end

	trials = NamedTuple[]
	rows = NamedTuple[]
	for threshold in options.thresholds
		threshold_trials = [
			run_schelling_trial(config_for(threshold, options.seed + offset), options.steps, options.mode)
			for offset in 0:options.seeds-1
		]
		append!(trials, threshold_trials)
		push!(rows, (
			threshold,
			mean_satisfied = mean(row.satisfied for row in threshold_trials),
			mean_segregation = mean(row.segregation for row in threshold_trials),
			mean_last_swapped = mean(row.last_swapped for row in threshold_trials),
			mean_steps = mean(row.steps_run for row in threshold_trials),
			converged_runs = count(row -> row.converged, threshold_trials),
		))
	end

	total_ticks = sum(row.steps_run for row in trials)
	total_elapsed_ns = sum(row.elapsed_ns for row in trials)
	return (
		options = options,
		trials = trials,
		rows = rows,
		total_ticks = total_ticks,
		total_elapsed_ns = total_elapsed_ns,
		mean_satisfied = mean(row.satisfied for row in trials),
		mean_segregation = mean(row.segregation for row in trials),
		mean_last_swapped = mean(row.last_swapped for row in trials),
		actual_steps = mean(row.steps_run for row in trials),
	)
end

function write_schelling_study_csv(result; io = stdout)
	println(io, "threshold,mean_satisfied_pct,mean_segregation_index,mean_last_swapped,mean_steps,converged_runs")
	for row in result.rows
		@printf(io, "%.2f,%.4f,%.4f,%.2f,%.2f,%d\n", row.threshold, row.mean_satisfied, row.mean_segregation, row.mean_last_swapped, row.mean_steps, row.converged_runs)
	end
	elapsed_ms = result.total_elapsed_ns / 1_000_000
	tpms = result.total_elapsed_ns == 0 ? 0.0 : result.total_ticks / elapsed_ms
	mspt = result.total_ticks == 0 ? 0.0 : elapsed_ms / result.total_ticks
	println(io, "performance_metric,total_ticks,elapsed_ms,tpms,mspt")
	@printf(io, "performance,%d,%.3f,%.6f,%.6f\n", result.total_ticks, elapsed_ms, tpms, mspt)
	return nothing
end
